lib/pbench/server/api/resources/query_apis/datasets_publish.py

```diff
@@ -0,0 +1,207 @@
+        {"authorization.access": "public"}\n
+
+        We rely on the document ID map built by the indexer to find the many
+        Elasticsearch documents created by Pbench indexing. This avoids the
+        need to use multiple Elasticsearch queries here to determine the set
+        of documents on which to act.
+
+        TODO: If some of the bulk updates fail, we need a mechanism either to
+        RETRY or to ROLL BACK (which would in turn require a RETRY); TBD.
+        """
+        user = json_data["user"]
+        name = json_data["name"]
+        access = json_data["access"]
+
+        self.logger.info(
+            "Update access for dataset {} for user {} to {}, prefix {}",
+            name,
+            user,
+            access,
+            self.prefix,
+        )
+
+        map = json.loads(Metadata.get(dataset=context, key=Metadata.INDEX_MAP).value)
+
+        # Construct an "NDJSON" document for a bulk update of all Elasticsearch
+        # documents associated with the Dataset.
+        #
+        # NDJSON is a series of JSON documents separated by "newlines" (defined
+        # by Elasticsearch as cr-lf pairs). We'll construct stringified JSON in
+        # an array which is then gathered up using str.join().
+        #
+        # TODO: it'd probably be better to use the Elasticsearch bulk helper and
+        # pyesbulk here; but I don't want to duplicate the setup in indexer.py,
+        # it's not well generalized, and I don't want to add that refactoring to
+        # this PR.
+
+        cmds = []
+        for index, ids in map.items():
+            for id in ids:
+                x = (
+                    f'{{"update": {{"_index": "{index}", "_id": "{id}"}}}}\n'
+                    f'{{"doc": {{"authorization": {{"access": "{access}"}}}}}}'
+                )
+                cmds.append(x)
+        data = "\n".join(cmds) + "\n"
+        return {
+            "path": "/_bulk",
+            "kwargs": {
+                "data": data,
+                "headers": {"Content-Type": "application/x-ndjson"},
+                "params": {"refresh": "true"},
+            },
+        }
+
+    def postprocess(
+        self, es_json: Dict[AnyStr, Any], context: Dataset
+    ) -> Dict[AnyStr, Any]:
+        """
+        A bulk update succeeds if any of the individual updates succeeds, and
+        PUBLISH can affect hundreds of documents across many Elasticsearch
+        indices. Here we'll look for the "errors" boolean in the response
+        payload, which is true if any failures occurred. We'll then scan the
+        "items" list to diagnose the specific failures in the log.
+
+        FIXME: What do we do about error recovery here? At this point we may
+        have "published" some subset of the Elasticsearch documents but not
+        others. Do we try to "unpublish" those failed documents and fail the
+        whole operation? (And what if some of the "unpublish" operations fail
+        as well?)
```

Comment:
Yeah, I'm not updating the `Dataset` access except on success; so a re-publish ought to try again. (The map doesn't change, and it's not sensitive to the current state of the document field in Elasticsearch.)

There's no harm in re-publishing documents that were already published, since we're just setting "authorization.access" to "public" (or I suppose to "private"). It's idempotent. Of course, if you change your mind and try to set everything back to the original state, nothing will happen because it matches the current `Dataset` state. I'm not sure I want to track a third state to deal with that, however.

Reply:
I'm thinking that, until we have more experience with this (or until we institute a third state), we should omit the optimization and just attempt whatever operation the user requested.  As you say, it's idempotent, so is there really a significant downside to re-publishing already published data (or "re-privating" it)?